### ISSUE TITLE
set_ticks() on shared axis

### ISSUE THREAD

**Issue by avdeevvadim** [matplotlib/matplotlib#8320]
### Bug report

**Bug summary**

Not sure it's a bug, but in the following example the order in which you change subplots properties matters.
Suppose you have several subplots with the shared axis and you set ticks for this axis with the range outside the data. Doing it for all subplots one by one, you get ticks placed correctly on the last subplot and incorrectly on all others.

**Code for reproduction**

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(3, 1, sharex=True)

ax[0].scatter([0, 1], [0, 1])
ax[0].set_xticks([0, 1, 2])

ax[1].scatter([0, 1], [0, 1])
ax[1].set_xticks([0, 1, 2])

ax[2].scatter([0, 1], [0, 1])
ax[2].set_xticks([0, 1, 2])

plt.show()
```

**Actual outcome**
![matplotlib_wrong](https://cloud.githubusercontent.com/assets/13047840/24057209/acfdeba0-0b57-11e7-9fff-90557a0c986e.png)

**Expected outcome**

Let's try the same code, but change all subplots properties afterwards.

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(3, 1, sharex=True)

ax[0].scatter([0, 1], [0, 1])

ax[1].scatter([0, 1], [0, 1])

ax[2].scatter([0, 1], [0, 1])

plt.setp(ax, xticks=[0, 1, 2])

"""
Alternatively,
ax[0].set_xticks([0, 1, 2])
ax[1].set_xticks([0, 1, 2])
ax[2].set_xticks([0, 1, 2])
"""

plt.show()
```

Result:
![matplotlib_right](https://cloud.githubusercontent.com/assets/13047840/24057210/acffd7da-0b57-11e7-86bc-10affb035eb2.png)

**Not so expected outcome**

Finally, change all subplots properties somewhere earlier.

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(3, 1, sharex=True)

ax[0].scatter([0, 1], [0, 1])

plt.setp(ax, xticks=[0, 1, 2])

ax[1].scatter([0, 1], [0, 1])

ax[2].scatter([0, 1], [0, 1])

plt.show()
```

Result:
![matplotlib_not_expected](https://cloud.githubusercontent.com/assets/13047840/24059639/dd972746-0b60-11e7-98ed-96a55b339f86.png)

**Matplotlib version**
Matplotlib 2.0.0 (installed using Anaconda)
Python 3.5.3
Fedora 25

**Comment by HHest** [comment]
I think when one performs set_xticks on one Axes object, the others should follow since sharex=True. 
If this rule is respected, then the sequencing issue would go away.

In other words:

```
    fig, ax = plt.subplots(3, 1, sharex=True)
    ax[0].scatter([0, 1], [0, 1])
    ax[1].scatter([0, 1], [0, 1])
    ax[2].scatter([0, 1], [0, 1])
    ax[0].set_xticks([0, 1, 2])
    plt.show()
```

should produce:
![image](https://user-images.githubusercontent.com/3169669/47078045-af713680-d202-11e8-8fd4-d1978cd0ef6f.png)

Currently it is producing:
![image](https://user-images.githubusercontent.com/3169669/47078173-f95a1c80-d202-11e8-80f3-2e3ada1836a5.png)

**Comment by csbrown** [comment]
Is there a workaround for this?

**Comment by ImportanceOfBeingErnest** [comment]
@csbrown Yes, the workaround is found under the headline "Expected outcome" in the original post.

**Comment by QuLogic** [comment]
This was fixed by #18529.
